hw10/hw10_t01.py

```diff
@@ -0,0 +1,212 @@
+"""
+Ваша задача спарсить информацию о компаниях, находящихся в индексе S&P 500 с данного сайта.
+
+https://markets.businessinsider.com/index/components/s&p_500
+
+Для каждой компании собрать следующую информацию:
+
+    Текущая стоимость в рублях (конвертацию производить по текущему курсу, взятому с сайта центробанка РФ)
+    Код компании (справа от названия компании на странице компании)
+    P/E компании (информация находится справа от графика на странице компании)
+    Годовой рост/падение компании в процентах (основная таблица)
+    Высчитать какую прибыль принесли бы акции компании (в процентах), если бы они были куплены на уровне 52 Week Low и проданы на уровне 52 Week High (справа от графика на странице компании)
+
+Сохранить итоговую информацию в 4 JSON файла:
+
+    Топ 10 компаний с самими дорогими акциями в рублях.
+    Топ 10 компаний с самым низким показателем P/E.
+    Топ 10 компаний, которые показали самый высокий рост за последний год
+    Топ 10 комппаний, которые принесли бы наибольшую прибыль, если бы были куплены на самом минимуме и проданы на самом максимуме за последний год.
+    Пример формата:
+
+[
+{
+    "code": "MMM",
+    "name": "3M CO.",
+    "price" | "P/E" | "growth" | "potential profit" : value,
+},
+...
+]
+"""
+import asyncio
+import datetime as dt
+import json
+import re
+from typing import Iterator
+
+import aiohttp
+from bs4 import BeautifulSoup
+
+INF = float("INF")
+NEG_INF = -float("INF")
+company_dummy = {
+    "name": "Dummy",
+    "code": "DUMMY",
+    "price": NEG_INF,
+    "P/E": INF,
+    "growth": NEG_INF,
+    "potential profit": NEG_INF,
+}
+highest_price = [company_dummy] * 10
+lowest_pe = [company_dummy] * 10
+highest_growth = [company_dummy] * 10
+highest_potential_profit = [company_dummy] * 10
+
+
+async def get_sp_500_info() -> None:
+    base_url = "https://markets.businessinsider.com/index/components/s&p_500"
+    base_company_url = "https://markets.businessinsider.com"
+    async with aiohttp.ClientSession() as session:
+
+        usd = await get_usd_currency_from_cbr(session)
+
+        for page in range(1, 13):
+            async with session.get(base_url, params=[("p", page)]) as resp:
+
+                html = await resp.text()
+                soup = BeautifulSoup(html, features="html.parser")
+
+                for company in get_all_stocks_from_page(soup):
+                    name = company["title"]
+                    growth = float(
+                        list(list(company.parent.next_siblings)[-4].children)[
+                            -2
+                        ].string.rstrip("%")
+                    )
+
+                    company_url = base_company_url + company["href"]
+                    asyncio.create_task(
+                        get_company_info(company_url, usd, name, growth)
+                    )
+                    await asyncio.sleep(0.05)
```

Comment:
Why?

Reply:
I suppose the server is cutting me off after 350-380 requests. I decided to slow down this way. After that, I've been able to collect all data.

Reply:
This knowledge should be in the code.